src/java/org/apache/cassandra/cql3/CqlConstraint.java

```diff
@@ -0,0 +1,145 @@
+
+        public CqlConstraint prepareWithName(ColumnIdentifier constraintName)
+        {
+            return new CqlConstraint(constraintName, null, constraintCondition);
+        }
+    }
+
+    public CqlConstraint(ColumnIdentifier constraintName, ColumnIdentifier columnName, ConstraintCondition constraintCondition)
+    {
+        if (constraintName == null)
+        {
+            final String randomConstraintName = UUID.randomUUID().toString().replace("-", "");
+            this.constraintName = new ColumnIdentifier(randomConstraintName, false);
+        }
+        else
+        {
+            this.constraintName = constraintName;
+        }
+        this.columnName = columnName;
+        this.constraintCondition = constraintCondition;
+    }
+
+    public void appendCqlTo(CqlBuilder builder)
+    {
+        builder.append(" CHECK ").append(toString());
+    }
+
+    public void checkConstraint(Map<String, String> columnValues, ColumnMetadata columnMetadata, TableMetadata tableMetadata)
+    {
+        constraintCondition.checkCondition(columnValues, columnMetadata, tableMetadata);
+    }
+
+    public void validateConstraint(ColumnMetadata columnMetadata, TableMetadata tableMetadata)
+    {
+        constraintCondition.validateCondition(columnMetadata, tableMetadata);
+    }
+
+    @Override
+    public String toString()
+    {
+        return constraintCondition.toString();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(constraintName);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        return obj instanceof CqlConstraint
+               && Objects.equal(constraintName, ((CqlConstraint) obj).constraintName)
+               && Objects.equal(constraintCondition, ((CqlConstraint) obj).constraintCondition);
+    }
+
+
+    public static class Serializer
+    {
+        public void serialize(CqlConstraint cqlConstraint, DataOutputPlus out, Version version) throws IOException
+        {
+            out.writeUTF(cqlConstraint.constraintName.toString());
+            out.writeUTF(cqlConstraint.columnName.toString());
+            out.writeUTF(cqlConstraint.constraintCondition.getClass().toString());
+            cqlConstraint.constraintCondition.serialize(cqlConstraint.constraintCondition, out, version);
+        }
+
+        public CqlConstraint deserialize(DataInputPlus in, String keyspace, AbstractType<?> columnType, Types types, UserFunctions functions, Version version) throws IOException
+        {
+            String nameText = in.readUTF();
+            String columnName = in.readUTF();
+            ColumnIdentifier identifier = new ColumnIdentifier(nameText, true);
+            ColumnIdentifier columnNameIdentifier = new ColumnIdentifier(columnName, true);
+            String columnConstraintClassName = in.readUTF();
+            ConstraintCondition condition;
+            if (columnConstraintClassName.equals(CqlConstraintFunctionCondition.class.getName()))
```

Comment:
are we sure that everything `columnConstraintClassName` will ever be is either `CqlConstraintFunctionCondition` or `ConstraintScalarCondition`? Or in other words, are we sure that if it is not condition then it is scalar condition? That is quite an assumption ...